presto-main/src/main/java/com/facebook/presto/sql/expressions/ExpressionOptimizerManager.java

```diff
@@ -0,0 +1,103 @@
+    public static final String EXPRESSION_MANAGER_FACTORY_NAME = "expression-manager-factory.name";
+
+    private final Map<String, ExpressionOptimizerFactory> expressionOptimizerFactories = new ConcurrentHashMap<>();
+    private final AtomicReference<ExpressionOptimizer> rowExpressionInterpreter = new AtomicReference<>();
+    private final NodeManager nodeManager;
+    private final FunctionAndTypeManager functionAndTypeManager;
+    private final FunctionResolution functionResolution;
+    private final ExpressionOptimizer defaultExpressionOptimizer;
+
+    @Inject
+    public ExpressionOptimizerManager(PluginNodeManager nodeManager, FunctionAndTypeManager functionAndTypeManager)
+    {
+        requireNonNull(nodeManager, "nodeManager is null");
+        this.nodeManager = requireNonNull(nodeManager, "nodeManager is null");
+        this.functionAndTypeManager = requireNonNull(functionAndTypeManager, "functionAndTypeManager is null");
+        this.functionResolution = new FunctionResolution(functionAndTypeManager.getFunctionAndTypeResolver());
+        this.defaultExpressionOptimizer = new RowExpressionOptimizer(functionAndTypeManager);
+        rowExpressionInterpreter.set(defaultExpressionOptimizer);
+    }
+
+    public void loadExpressionOptimizerFactory()
+    {
+        try {
+            if (EXPRESSION_MANAGER_CONFIGURATION.exists()) {
+                Map<String, String> properties = new HashMap<>(loadProperties(EXPRESSION_MANAGER_CONFIGURATION));
+                loadExpressionOptimizerFactory(properties);
+            }
+        }
+        catch (IOException e) {
+            throw new UncheckedIOException("Failed to load expression manager configuration", e);
+        }
+    }
+
+    private void loadExpressionOptimizerFactory(Map<String, String> properties)
+    {
+        properties = new HashMap<>(properties);
+        String factoryName = properties.remove(EXPRESSION_MANAGER_FACTORY_NAME);
+        checkArgument(!isNullOrEmpty(factoryName), "%s does not contain %s", EXPRESSION_MANAGER_CONFIGURATION, EXPRESSION_MANAGER_FACTORY_NAME);
+        checkArgument(
+                rowExpressionInterpreter.compareAndSet(
+                        defaultExpressionOptimizer,
+                                expressionOptimizerFactories.get(factoryName).createOptimizer(properties, new ExpressionOptimizerContext(nodeManager, functionAndTypeManager, functionResolution))),
+                "ExpressionManager is already loaded");
+    }
+
+    public void addExpressionOptimizerFactory(ExpressionOptimizerFactory expressionOptimizerFactory)
+    {
+        String name = expressionOptimizerFactory.getName();
+        checkArgument(
+                this.expressionOptimizerFactories.putIfAbsent(name, expressionOptimizerFactory) == null,
+                "ExpressionOptimizerFactory %s is already registered", name);
+    }
+
+    @Override
+    public ExpressionOptimizer getExpressionOptimizer()
```

Comment:
A thought I had while looking at the structure of the expression optimizer manager:

```suggestion
    public ExpressionOptimizer getExpressionOptimizer(Session session)
```


Do you think it would be beneficial to add the ability to pass session properties here? I could imagine a case where expression optimization could be configured in the future by the session or potentially switch/update configurations by the end user.

The current implementation requires the user to update config files and restart the server if there were other configs outside of just enabling/disabling the delegating optimizer. It might also allow you to encapsulate the configuration logic for returning the `DelegatingExpressionOptimizer` entirely in this class rather than having the current conditional inside of the `SimplifyRowExpressions` class.

Reply:
I've been thinking about this for other plugins as well- of having the ability to pass session properties to plugins. Right now we only support session properties for connectors, but we should support them for other plugins too.  When I looked into this before, I think we needed to basically create a connectorSession for these plugins as well and be able to get session properties based on the name prefix of the plugin.  The only thing is that it means you can't e.g. give your catalog and function namespace manager the same name.

Reply:
I think I can get this to work for now by passing in a ConnectorSession, and casting to FullConnectorSession inside of ExpressionOptimizerManager.  I agree this will reduce the branching that both of you have pointed out.  I agree we should have a design around plugin sessions as well.

Reply:
Yes, this will work.  Thanks for the suggestion @ZacBlanco.

It can't be a `Session` because these optimizers are used in connectors as well, and those connectors won't have access to `Session` (see `RowExpressionService`).  But we can pass in a `ConnectorSession`, and convert it back to a `FullConnectorSession` as a compromise.

This eliminates the branch in `SimplifyRowExpressions` and centralizes where one toggles the expression evaluation code.  I think it's cleaner this way in spite of the hack to convert `ConnectorSession` to `FullConnectorSession`.  It also guarantees that you use the right `ExpressionOptimizer` as long as you retrieve it from the appropriate interfaces (`RowExpressionService` in a connector, `ExpressionOptimizerManager` in the planner).  Let me know what you think of this approach.